# Clase 2 — Tipos de Datos, Operadores y Modelo de Objetos

Bienvenido a la segunda clase del curso de Python. En esta sesión exploraremos el modelo de objetos, los tipos básicos, los operadores y cómo manejar cadenas de texto en profundidad.

**Contenido de esta clase:**
1. Extensión de archivos y Punto de Entrada
2. La función `print()`
3. Tipado dinámico y Modelo de Objetos
4. Palabras Reservadas
5. Tipos de Datos Primitivos (en profundidad)
6. Conversión de Tipos
7. Operadores (Aritméticos, Comparación, Lógicos, Bits, Asignación)
8. Cadenas de Texto (Strings) en profundidad
9. Ejercicio Práctico: Calculadora de Calificaciones

---
## 1. Extensión de archivos y Punto de Entrada

Los scripts en Python se guardan con la extensión **`.py`**. Son archivos de texto que son interpretados por el intérprete de Python (Corcuera, IntroPython).

Cuando ejecutas un archivo directamente (por ejemplo: `python script.py`), Python asigna un nombre especial a ese módulo: `"__main__"`.

Esto nos permite definir un **punto de entrada** explícito, similar a la función `main()` en C.

> **Puente con C:** En C, el punto de entrada es siempre la función `main()`. En Python, el código se ejecuta de arriba a abajo, pero la convención `if __name__ == '__main__':` nos permite controlar qué código se ejecuta solo cuando el archivo es el programa principal (y no cuando es importado como módulo).

In [1]:
# mi_modulo.py
def main():
    print("¡Este es el punto de entrada principal! Solo se ejecutará si este archivo se ejecuta directamente.")

print("¡Este código se ejecutará incluso si se importa como módulo!")

if __name__ == '__main__':
    main()

¡Este código se ejecutará incluso si se importa como módulo!
¡Este es el punto de entrada principal! Solo se ejecutará si este archivo se ejecuta directamente.


---
## 2. La función `print()`

Antes de sumergirnos en tipos y operadores, conozcamos la herramienta que usaremos para ver resultados: la función `print()`.

`print()` es una **función predefinida** (built-in) que envía texto a la salida estándar (la consola). Por defecto, a diferencia de C, python importa automáticamente la librería estándar. Sus características más comunes son:

| Uso | Ejemplo | Resultado |
|-----|---------|-----------|
| **Un solo valor** (agrega el salto de línea por defecto) | `print("Hola")` | `Hola` |
| **Varios valores** (separados por espacio) | `print("Hola", "Mundo")` | `Hola Mundo` |
| **Parámetro `sep`** (cambia el separador) | `print("a", "b", "c", sep="-")` | `a-b-c` |
| **Parámetro `end`** (cambia el final, por defecto `\n`) | `print("Hola", end=" ")` + `print("Mundo")` | `Hola Mundo` |
| **Formato con f-strings** | `print(f"2 + 2 = {2 + 2}")` | `2 + 2 = 4` |
| **Representación con `!r`** | `print(f"{'Hola'!r}")` | `'Hola'` |

> **Nota:** En una clase futura sobre funciones exploraremos `print()` a fondo y veremos todas las funciones predefinidas. Por ahora, con estos usos básicos tenemos suficiente para lo que viene.

Ejecuta la celda de abajo para ver `print()` en acción:

In [2]:
print("Hola")                          # Un solo valor
print("Hola", "Mundo")               # Varios valores
print("a", "b", "c", sep="-")      # Separador personalizado
print("Hola", end=" ")               # Sin salto de línea al final
print("Mundo")                        # Continúa en la misma línea
print(f"2 + 2 = {2 + 2}")            # f-string
print(f"{'Python'!r}")               # Representación con comillas

Hola
Hola Mundo
a-b-c
Hola Mundo
2 + 2 = 4
'Python'


---
## 3. Tipado dinámico y Modelo de Objetos

### Todo es un Objeto

En Python, **todo es un objeto** (incluso las funciones y los números). Cada objeto se caracteriza por tres propiedades fundamentales (Corcuera, IntroPython):

- **Identidad**: un identificador único en memoria (obtenible con `id()`).
- **Tipo**: la clase del objeto (obtenible con `type()`).
- **Valor**: el dato que almacena.

### Tipado Dinámico

Python tiene **tipado dinámico**, lo que significa que:
- El tipo está asociado al **valor (objeto)** en memoria, **no a la variable** en sí.
- No necesitas declarar el tipo de una variable antes de usarla.
- Una variable se crea automáticamente cuando un valor es asignado la primera vez.
- Una variable puede reasignarse a objetos de cualquier tipo.

> **Puente con C:** A diferencia de C, donde una variable reserva un espacio de memoria con un tamaño fijo para un tipo específico (`int x = 5;`), en Python **todas las variables actúan como punteros (referencias)** a objetos en memoria. No existen tipos primitivos "por valor". El recolector de basura (Garbage Collector) se encarga de liberar la memoria cuando un objeto ya no tiene referencias.

> **Concepto clave (MoureDev, Guía Junior 2026):** Python usa **tipado dinámico** (tipos verificados durante la ejecución) y **tipado fuerte** (conversiones de tipo deben ser explícitas). Esto contrasta con C que es de tipado estático (tipos verificados en compilación).

In [3]:
# Ejemplo de tipado dinámico, los errores se detectan en tiempo de ejecución (runtime)
x = 1
print(f"x = {x}, \t\ttipo = {type(x)}") # \t es un tabulador para alinear la salida

x = 1.0
print(f"x = {x}, \ttipo = {type(x)}")

x = 'hello'
print(f"x = {x}, \ttipo = {type(x)}")

x = [1, 2, 3]
print(f"x = {x}, \ttipo = {type(x)}")

# Ejemplo de tipado fuerte, el complilador o el intérprete no permite operaciones entre tipos incompatibles
# x = 1 + '1'  # Esto generará un error de tipo (TypeError) en tiempo de ejecución


x = 1, 		tipo = <class 'int'>
x = 1.0, 	tipo = <class 'float'>
x = hello, 	tipo = <class 'str'>
x = [1, 2, 3], 	tipo = <class 'list'>


### Identidad de objetos: `id()` y el operador `is`

La función `id()` devuelve la identidad única de un objeto. El operador `is` compara si dos variables apuntan al **mismo objeto** en memoria (no solo si tienen el mismo valor).

---
## 4. Palabras Reservadas

Al igual que en C, existen palabras clave que no puedes usar como identificadores (nombres de variables o funciones). Las palabras reservadas en Python 3 son (Corcuera, IntroPython):

| | | | | | |
|---|---|---|---|---|---|
| `False` | `None` | `True` | `and` | `as` | `assert` |
| `async` | `await` | `break` | `class` | `continue` | `def` |
| `del` | `elif` | `else` | `except` | `finally` | `for` |
| `from` | `global` | `if` | `import` | `in` | `is` |
| `lambda` | `nonlocal` | `not` | `or` | `pass` | `raise` |
| `return` | `try` | `while` | `with` | `yield` | |

Aunque muchas de ellas las estudiaremos en profundidad más adelante, conviene tener claro desde el inicio **qué hace cada una, en qué contexto se usa y cuál es la mejor manera de utilizarla**.

Para que las explicaciones no se sientan abstractas, vamos a usar un **mismo contexto narrativo a lo largo de toda la sección**: imaginemos que estamos construyendo un **sistema de gestión de calificaciones**. Cada keyword aparecerá en el momento del flujo donde naturalmente se usaría en un proyecto real:

1. Empezamos importando módulos de la librería estándar.
2. Definimos funciones y constantes base.
3. Validamos los datos de entrada del usuario.
4. Filtramos y procesamos las calificaciones.
5. Manejamos errores gracefully.
6. Guardamos un reporte en disco.

Verás cómo las keywords se encadenan como piezas de un rompecabezas.

### El operador asignación (`=`)

En Python **no es necesario declarar las variables** antes de usarlas. La asignación inicial crea la variable y enlaza el valor a la variable (Corcuera, IntroPython).

Python soporta varias formas de asignación:
- **Asignación simple:** `x = 5`
- **Asignación múltiple (por pares):** `b, c, d = 123, 4.5, 'Hello'`
- **Asignación en cadena:** `e = f = g = 123`
- **Eliminación:** `del x` elimina la variable

### 4.1 Importación de módulos: `import`, `from`, `as`

Nuestro sistema de calificaciones necesitará generar reportes, redondear números y quizá leer/escribir archivos. Nada de eso lo construiremos desde cero: **importaremos módulos de la biblioteca estándar**.

**`import`** — Carga un módulo completo bajo su nombre.
- **Qué hace:** Hace disponible un módulo entero en el espacio de nombres.
- **Frecuencia de uso:** **Extremadamente alta**.
- **Manera correcta de uso:** Una línea por módulo, en la parte superior del archivo.
- **Consejo:** Evita `import *` (trae todo sin control). Mejor ser explícito (PEP 20).

```python
import statistics  # Para calcular promedios y medianas
import math        # Para redondear y funciones matemáticas
import os          # Para interactuar con el sistema de archivos
```

**`from ... import`** — Importa elementos específicos.
- **Qué hace:** Trae solo lo que necesitas al espacio de nombres actual, sin prefijo.
- **Contexto:** Cuando solo usarás unas pocas funciones de un módulo.
- **Consejo:** Cuidado con colisiones: si haces `from math import pi` y defines tu propia variable `pi`, tendrás un conflicto.

```python
from statistics import mean, median
from math import ceil
```

**`as`** — Crea un alias al importar.
- **Qué hace:** Renombra el módulo importado.
- **Contexto:** Nombres largos, conflictos con variables locales, o seguir convenciones de la comunidad (`numpy as np`, `pandas as pd`).

```python
import statistics as stats
```

### 4.2 Definición de funciones: `def`, `return`, `lambda`, `yield`

Una vez importados los módulos, necesitamos **definir la lógica de negocio** de nuestro sistema.

**`def`** — Define una función reusable.
- **Qué hace:** Crea un bloque de código con nombre que podemos llamar múltiples veces.
- **Frecuencia de uso:** **Extremadamente alta**. Es el bloque de construcción de la modularidad.
- **Manera correcta de uso:** Nombres en `snake_case`, verbos descriptivos, docstring opcional.
- **Consejo:** Una buena función hace **una sola cosa**. Si pasa de 20-30 líneas, probablemente deba dividirse.

**`return`** — Devuelve un valor desde la función.
- **Qué hace:** Termina la ejecución de la función y opcionalmente devuelve un resultado.
- **Frecuencia de uso:** **Extremadamente alta**.
- **Manera correcta de uso:** Si la función no tiene `return` explícito, retorna `None` implícitamente.
- **Consejo:** "Guard Clauses": maneja primero los casos de error y retorna temprano. Más legible que `if/else` anidados.

```python
def calcular_promedio_ponderado(notas, pesos):
    """Calcula la nota final del semestre usando promedio ponderado.

    Args:
        notas: lista de calificaciones (0-10).
        pesos: lista de créditos/pesos de cada materia.

    Returns:
        Promedio ponderado redondeado a 2 decimales, o None si hay error.
    """
    if len(notas) != len(pesos):
        return None  # Guard clause: entrada inválida
    if sum(pesos) == 0:
        return None  # Guard clause: pesos inválidos

    promedio = sum(n * p for n, p in zip(notas, pesos)) / sum(pesos)
    return round(promedio, 2)
```

**`lambda`** — Función anónima de una sola expresión.
- **Qué hace:** Crea una función sin nombre, en una sola línea.
- **Frecuencia de uso:** **Media**. Útil pero no abusar.
- **Contexto:** Como argumento de funciones de orden superior (`map`, `filter`, `sorted`, `max`).
- **Consejo:** Si la lógica no cabe en una expresión clara, usa `def`. Las lambdas ilegibles son una mala práctica.

```python
# Necesitamos clasificar cada nota. lambda es perfecta aquí:
clasificar = lambda nota: "Aprobado" if nota >= 7 else "Reprobado"
print(clasificar(8.5))  # Aprobado
```

**`yield`** — Produce un valor en una función generadora.
- **Qué hace:** Convierte la función en un generador que produce valores perezosamente (uno a la vez, bajo demanda).
- **Frecuencia de uso:** **Media**.
- **Contexto:** Procesar listas enormes sin cargarlas todas en memoria, secuencias infinitas, pipelines de datos.
- **Consejo:** Cada `yield` "pausa" la función hasta que alguien pida el siguiente valor. Ideal para reportes con miles de estudiantes.

```python
def generar_reporte(estudiantes):
    """Genera un reporte línea por línea sin cargar todo en memoria."""
    for est in estudiantes:
        yield f"{est['nombre']}: {est['promedio']:.2f}"
```

### 4.3 Control de flujo: `if`, `elif`, `else`, `for`, `while`, `break`, `continue`, `pass`

Ahora que tenemos las funciones base, necesitamos **dirigir el flujo del programa**: tomar decisiones y repetir tareas.

**`if` / `elif` / `else`** — Ejecución condicional.
- **Qué hacen:** `if` ejecuta un bloque si la condición es verdadera; `elif` añade condiciones alternativas; `else` cubre el caso por defecto.
- **Frecuencia de uso:** **Extremadamente alta**.
- **Manera correcta de uso:** Usa `elif` en vez de `if` anidados para mantener la legibilidad. Prefiere **guard clauses** cuando valides entradas.

```python
def obtener_calificacion_literal(promedio):
    """Convierte un promedio numérico a letra: A, B, C, D, F."""
    if promedio is None:
        return "N/A"           # Guard clause
    if promedio >= 9.0:
        return "A"
    elif promedio >= 8.0:
        return "B"
    elif promedio >= 7.0:
        return "C"
    elif promedio >= 6.0:
        return "D"
    else:
        return "F"             # Reprobado
```

**`for`** — Itera sobre una secuencia.
- **Qué hace:** Recorre los elementos de cualquier iterable (lista, string, diccionario, etc.).
- **Frecuencia de uso:** **Extremadamente alta**.
- **Puente con C:** El `for` de Python es como un `foreach`, no como el `for(int i=0; i<n; i++)` de C. Para iterar N veces, usa `for i in range(N):`.
- **Consejo:** Si el cuerpo del bucle tiene más de 3-4 líneas, probablemente convenga extraerlo a una función.

```python
# Procesar la lista de estudiantes para clasificarlos
estudiantes = [
    {"nombre": "Ana", "promedio": 9.2},
    {"nombre": "Luis", "promedio": 6.8},
    {"nombre": "María", "promedio": 8.5},
]

for estudiante in estudiantes:
    letra = obtener_calificacion_literal(estudiante["promedio"])
    print(f"{estudiante['nombre']}: {letra}")
```

**`while`** — Bucle mientras la condición sea verdadera.
- **Qué hace:** Repite un bloque mientras la condición sea `True`.
- **Frecuencia de uso:** **Alta**, pero menor que `for`.
- **Contexto:** Cuando **no sabes cuántas iteraciones** necesitas de antemano: validar entrada del usuario, esperar a que se cumpla una condición externa.
- **Consejo:** Asegúrate de que la condición eventualmente sea `False`, si no tendrás un bucle infinito.

```python
# Validar que el usuario ingrese una nota válida (entre 0 y 10)
def pedir_nota(materia):
    while True:  # Bucle hasta que el usuario ingrese algo válido
        entrada = input(f"Nota de {materia} (0-10): ")
        try:
            nota = float(entrada)
            if 0 <= nota <= 10:
                return nota        # Sale del bucle con return
            print("Debe estar entre 0 y 10.")
        except ValueError:
            print("Entrada inválida, intenta de nuevo.")
```

**`break`** — Sale del bucle inmediatamente.
- **Qué hace:** Termina el bucle antes de completar todas las iteraciones.
- **Contexto:** Encontrar un elemento, condición de salida temprana, evitar trabajo innecesario.

```python
# Buscar al primer estudiante con promedio perfecto
for estudiante in estudiantes:
    if estudiante["promedio"] == 10.0:
        print(f"¡{estudiante['nombre']} sacó 10!")
        break   # Ya lo encontramos, no seguir buscando
```

**`continue`** — Salta a la siguiente iteración.
- **Qué hace:** Ignora el resto del bloque y va directamente a la siguiente vuelta.
- **Contexto:** Filtrar elementos dentro de un bucle sin anidar `if`s.

```python
# Imprimir solo los estudiantes aprobados
for estudiante in estudiantes:
    if estudiante["promedio"] < 7.0:
        continue   # Saltar reprobados
    print(f"{estudiante['nombre']} aprobó con {estudiante['promedio']}")
```

**`pass`** — No hacer nada (placeholder).
- **Qué hace:** Es una operación nula. Sirve para cumplir la regla sintáctica de que un bloque no puede estar vacío.
- **Contexto:** Clases o funciones que aún no implementas, ramas `if` que dejarás para después.
- **Consejo:** Mejor `pass` con un `TODO` que dejar un bloque vacío (que causa `IndentationError`).

```python
class SistemaCalificaciones:
    def exportar_a_pdf(self):
        pass   # TODO: implementar exportación a PDF más adelante

    def exportar_a_csv(self):
        # Por ahora solo exportamos a CSV
        return "reporte.csv"
```

### 4.4 Operadores lógicos y de membresía: `and`, `or`, `not`, `in`, `is`

Para tomar decisiones, necesitamos **combinar y comparar** condiciones.

**`and`** — Conjunción lógica (Y).
- **Qué hace:** Retorna `True` solo si **ambos** operandos son verdaderos. Cortocircuito: si el primero es `False`, no evalúa el segundo.
- **Consejo:** Aprovecha el cortocircuito para evitar errores: `estudiante and estudiante["nombre"]` no falla si `estudiante` es `None`.

**`or`** — Disyunción lógica (O).
- **Qué hace:** Retorna `True` si al menos un operando es verdadero. Cortocircuito: si el primero es `True`, no evalúa el segundo.
- **Consejo:** Cuidado con `or` para valores por defecto cuando `0` o `""` son válidos (son falsy). En esos casos, usa un `if` explícito.

**`not`** — Negación lógica.
- **Qué hace:** Invierte el valor de verdad. Más legible que `== False`.

```python
def estudiante_destacado(est):
    """Determina si un estudiante es destacado: promedio >= 9 Y no tiene reprobadas."""
    if not est["materias"]:        # not: lista vacía es False
        return False
    if est["promedio"] >= 9.0 and not any(n < 7.0 for n in est["materias"]):
        return True
    return False
```

**`in` / `not in`** — Membresía en una secuencia.
- **Qué hace:** Verifica si un elemento está (o no) contenido en un iterable.
- **Contexto:** Búsquedas, validaciones, filtros.

```python
# Verificar que la materia esté en el catálogo permitido
catalogo = {"Cálculo", "Programación", "Física", "Química"}
materia = input("Materia: ") # input es una función para recibir un string del usuario

if materia not in catalogo:
    print(f"'{materia}' no está en el catálogo.")
else:
    print(f"'{materia}' es válida.")
```

**`is` / `is not`** — Comparación de identidad.
- **Qué hace:** Retorna `True` si ambas variables apuntan al **mismo objeto en memoria** (no solo al mismo valor).
- **Contexto:** Comparar con singletons: `None`, `True`, `False`.
- **Consejo:** **Nunca uses `is` para comparar valores** como números o strings. Usa `==`. La excepción son los singletons.

```python
# Patrón idiomático para verificar None
resultado = calcular_promedio_ponderado(notas, pesos)
if resultado is None:
    print("No se pudo calcular el promedio.")
else:
    print(f"Promedio: {resultado}")
```

### 4.5 Valores literales: `True`, `False`, `None`

Estos tres valores aparecen constantemente al trabajar con condiciones y datos faltantes.

**`True` y `False`** — Los dos únicos valores booleanos.
- **Frecuencia de uso:** **Extremadamente alta**.
- **Consejo:** Recuerda que `True == 1` y `False == 0` porque `bool` es subclase de `int`. Evita `x == True`; prefiere `if x:`.

**`None`** — Ausencia intencional de valor.
- **Qué hace:** Representa "nada" o "valor indefinido". Único valor del tipo `NoneType`.
- **Frecuencia de uso:** **Muy alta**.
- **Contexto:** Inicializar antes de calcular, por defecto de funciones, datos faltantes.
- **Consejo:** Compáralo **siempre con `is` / `is not`**, nunca con `==`.

```python
# None marca "aún no se calculó" o "no hay datos"
def buscar_estudiante(nombre, lista):
    """Retorna el estudiante o None si no existe."""
    for est in lista:
        if est["nombre"] == nombre:
            return est
    return None  # Explícitamente no encontrado

# Uso
ana = buscar_estudiante("Ana", estudiantes)
if ana is not None:           # 'is not None', no '!= None'
    print(f"Encontrada: {ana['promedio']}")
else:
    print("Estudiante no registrada.")
```

### 4.6 Manejo de errores: `try`, `except`, `finally`, `raise`, `assert`

El usuario puede ingresar cualquier cosa: letras en vez de números, notas negativas, pesos cero... Nuestro sistema debe **fallar con gracia**.

**`try` / `except`** — Captura excepciones.
- **Qué hace:** `try` ejecuta código riesgoso; `except` maneja el error si ocurre.
- **Frecuencia de uso:** **Muy alta**.
- **Consejo:** Captura **excepciones específicas** (`ValueError`, `FileNotFoundError`), nunca `except:` solo. Es la diferencia entre un sistema robusto y uno que oculta bugs.

**`finally`** — Se ejecuta siempre, haya error o no.
- **Contexto:** Limpieza de recursos que **deben** liberarse (cerrar archivos, conexiones a BD).

**`raise`** — Lanza una excepción manualmente.
- **Filosofía "Fail Fast":** Si algo está mal, falla **ahora** con un mensaje claro, en lugar de continuar con datos corruptos que generen errores更难 de debuggear más adelante.

**`assert`** — Verifica invariantes durante desarrollo.
- **Qué hace:** Lanza `AssertionError` si la condición es falsa. Se **desactiva** con `python -O`.
- **Consejo crítico:** **Nunca uses `assert` para validar entrada de usuario** (puede estar desactivado). Usa `raise` para validaciones reales. `assert` es para el programador, no para el usuario final.

```python
def registrar_calificacion(materia, nota_str, peso_str):
    """Registra una materia con su nota y peso. Valida y lanza errores claros."""
    # assert: invariantes que el programador garantiza (no entrada de usuario)
    assert isinstance(materia, str), "materia debe ser string"

    # raise: validación real de entrada del usuario
    try:
        nota = float(nota_str)
        peso = float(peso_str)
    except ValueError:
        raise ValueError(f"Nota o peso inválido: '{nota_str}', '{peso_str}'")

    if not 0 <= nota <= 10:
        raise ValueError(f"Nota {nota} fuera de rango [0, 10]")
    if peso <= 0:
        raise ValueError(f"Peso debe ser positivo, recibido: {peso}")

    return {"materia": materia, "nota": nota, "peso": peso}
```

Más adelante, cuando llamemos a `registrar_calificacion`, envolveremos la llamada en otro `try/except` para manejar el error de forma amigable al usuario (ver sección 4.7).

### 4.7 Administradores de contexto: `with`, `as`

Al final del flujo, querremos **guardar el reporte en un archivo**. En C tendrías que recordar llamar a `fclose()` o arriesgarte a perder datos. En Python, el `with` lo hace por ti.

**`with`** — Administrador de contexto.
- **Qué hace:** Garantiza que el código de limpieza se ejecute **automáticamente**, incluso si hay errores.
- **Frecuencia de uso:** **Alta** cuando trabajas con recursos externos.
- **Puente con C:** El `with` de Python reemplaza el patrón `fopen` + manual `fclose` + verificación de errores de C. Es **mucho más seguro**.

**`as`** (en `with` y `except`) — Asigna el recurso o la excepción a una variable.
- **Contexto:** Capturar el objeto para usarlo, o guardar los detalles del error.

```python
def guardar_reporte(contenido, nombre_archivo):
    """Guarda el reporte en un archivo de forma segura."""
    # El 'as archivo' asigna el recurso a la variable.
    # El 'with' garantiza que el archivo se cierre aunque ocurra un error.
    with open(nombre_archivo, "w", encoding="utf-8") as archivo:
        archivo.write(contenido)
    # Aquí el archivo YA está cerrado, incluso si hubo una excepción dentro.
```

Combinando lo que hemos visto hasta ahora, podemos imaginar cómo se ve el flujo completo:

```python
# 1. Importar (4.1)
import statistics

# 2. Definir funciones (4.2)
def calcular_promedio(notas):
    if not notas:    # guard clause
        return None
    return round(statistics.mean(notas), 2)

# 3. Validar entrada con manejo de errores (4.6, 4.7)
def pedir_nota(materia):
    while True:
        entrada = input(f"Nota de {materia}: ")
        try:
            nota = float(entrada)
        except ValueError as error:        # 'as' captura el error
            print(f"Entrada inválida: {error}")
            continue
        if 0 <= nota <= 10:
            return nota
        print("Debe estar entre 0 y 10.")

# 4. Procesar con control de flujo (4.3, 4.4)
notas = [pedir_nota(m) for m in ["Cálculo", "Programación", "Física"]]
promedio = calcular_promedio(notas)
if promedio is not None and promedio >= 7.0:    # 'is', 'and'
    print(f"Aprobado con {promedio}")

# 5. Guardar reporte (4.7)
with open("reporte.txt", "w") as f:
    f.write(f"Promedio: {promedio}")
```

### 4.8 Programación asíncrona: `async`, `await`

**`async`** — Declara una corrutina (función que puede pausarse y resumirse).
**`await`** — Pausa hasta que una corrutina termine y devuelve su resultado.
- **Frecuencia de uso:** **Baja-Media** (solo en código con mucha I/O).
- **Contexto:** Operaciones concurrentes: consultar varias APIs a la vez, descargar muchos archivos, servidores web.
- **Consejo:** Solo puedes usar `await` **dentro** de funciones declaradas con `async def`. En este curso no profundizaremos hasta la Clase 10.

```python
# Versión ASÍNCRONA: consultar 3 servicios en paralelo
import asyncio

async def consultar_servicio(nombre):
    await asyncio.sleep(1)   # Simula una consulta de red
    return f"{nombre}: OK"

async def revisar_sistemas():
    # Lanza las 3 consultas "en paralelo" (no se bloquea esperando una a la vez)
    resultados = await asyncio.gather(
        consultar_servicio("Calificaciones"),
        consultar_servicio("Asistencia"),
        consultar_servicio("Pagos"),
    )
    return resultados

# asyncio.run(revisar_sistemas())  # Tarda ~1s en vez de ~3s
```

### 4.9 Gestión de scope: `del`, `global`, `nonlocal`

**`del`** — Elimina una variable o atributo.
- **Qué hace:** Borra la referencia, permitiendo que el garbage collector libere memoria.
- **Frecuencia de uso:** **Baja-Media**.
- **Contexto:** Liberar memoria de estructuras grandes, limpiar estado.

**`global`** — Declara una variable como global dentro de una función.
- **Frecuencia de uso:** **Baja** (generalmente evitar).
- **Consejo:** Evítalo cuando puedas. El estado global dificulta el testing y el debugging. Si necesitas estado compartido, considera usar una clase.

**`nonlocal`** — Modifica una variable de una función envolvente (en closures).
- **Frecuencia de uso:** **Baja** (solo en closures y decoradores avanzados).

```python
# del: liberar memoria explícitamente
datos_historicos = [calculo for calculo in range(1_000_000)]
del datos_historicos   # Libera la memoria inmediatamente

# global: estado compartido (úsalo con moderación)
contador_reportes = 0

def generar_reporte():
    global contador_reportes   # Sin esto, crearía una variable local
    contador_reportes += 1
    return f"Reporte #{contador_reportes}"

# nonlocal: closures
def crear_contador():
    total = 0
    def incrementar():
        nonlocal total        # Modifica la 'total' de la función externa
        total += 1
        return total
    return incrementar

contador = crear_contador()
print(contador())   # 1
print(contador())   # 2
```

### 4.10 Otras keywords importantes: `class`, `del`, `pass`

**`class`** — Define una clase (POO).
- **Frecuencia de uso:** **Alta** en proyectos medianos/grandes.
- **Consejo:** No crees clases para todo. Usa diccionarios/listas cuando no necesites comportamiento. Una clase se justifica cuando encapsula **estado + comportamiento**.

```python
class Estudiante:
    """Representa a un estudiante con sus calificaciones."""
    def __init__(self, nombre):
        self.nombre = nombre
        self.calificaciones = []    # estado

    def agregar_calificacion(self, materia, nota, peso):    # comportamiento
        self.calificaciones.append({
            "materia": materia, "nota": nota, "peso": peso
        })

    def calcular_promedio(self):
        if not self.calificaciones:
            return None
        total = sum(c["nota"] * c["peso"] for c in self.calificaciones)
        pesos = sum(c["peso"] for c in self.calificaciones)
        return round(total / pesos, 2)
```

**`pass`** — No hacer nada. Útil como placeholder:

```python
class SistemaReportes:
    def exportar_pdf(self):
        pass   # TODO: pendiente de implementar

    def exportar_csv(self):
        return "reporte.csv"
```

### 4.11 Resumen visual: el flujo de nuestro sistema de calificaciones

Para cerrar, observa cómo las keywords que vimos se conectan en un solo programa. **No te preocupes si no entiendes cada línea**; lo importante es ver el mapa general:

```python
import statistics as stats
from math import ceil

def pedir_nota(materia):
    while True:
        try:
            nota = float(input(f"Nota de {materia}: "))
            if 0 <= nota <= 10:
                return nota
        except ValueError:
            pass   # si no es número, vuelve a pedir
        print("Inválida, intenta de nuevo.")

def calcular_promedio(notas):
    if not notas:
        return None
    return round(stats.mean(notas), 2)

def obtener_letra(promedio):
    if promedio is None:
        return "N/A"
    if promedio >= 9: return "A"
    elif promedio >= 8: return "B"
    elif promedio >= 7: return "C"
    else: return "F"

# --- Flujo principal ---
materias = ["Cálculo", "Programación", "Física"]
notas = [pedir_nota(m) for m in materias]
promedio = calcular_promedio(notas)
letra = obtener_letra(promedio)

if promedio is not None and promedio >= 7.0:
    print(f"¡Aprobado! Promedio: {promedio} ({letra})")
else:
    print(f"Reprobado. Promedio: {promedio}")

with open("reporte.txt", "w") as f:
    f.write(f"Promedio final: {promedio}")
```

**¿Cuántas keywords reconoces?** Probablemente casi todas. Las iremos dominando una por una a lo largo del curso. Por ahora, quédate con la idea de que **cada keyword tiene un propósito claro y aparece en un momento específico del flujo**.

In [4]:
import keyword
print(f"Total de palabras reservadas: {len(keyword.kwlist)}")
print(f"Todas las keywords: {keyword.kwlist}")
print()

# Verificar si una palabra es reservada
print(f"¿'for' es reservada? {keyword.iskeyword('for')}")
print(f"¿'variable' es reservada? {keyword.iskeyword('variable')}")
print(f"¿'class' es reservada? {keyword.iskeyword('class')}")

# Intentar usar una keyword como variable genera SyntaxError
# Descomenta la siguiente línea para ver el error:
# for = 5  # SyntaxError: invalid syntax
print("\nNo puedes usar palabras reservadas como nombres de variables.")

# --- Ejemplo unificado: mini sistema de calificaciones ---
# Demuestra varias keywords trabajando juntas (sin pedir input al usuario).
print("\n--- Demo: sistema de calificaciones (sin input) ---\n")

# Import (4.1)
import statistics as stats

# def, return, guard clauses (4.2)
def calcular_promedio(notas):
    if not notas:           # if, not, in implícito
        return None        # return temprano (guard clause)
    return round(stats.mean(notas), 2)

# class (4.10)
class Estudiante:
    def __init__(self, nombre, calificaciones):
        self.nombre = nombre
        self.calificaciones = calificaciones

    def promedio(self):
        return calcular_promedio(self.calificaciones)

    def aprobado(self):
        # and, is not, comparación
        return self.promedio() is not None and self.promedio() >= 7.0

# for, if/elif/else, in, f-strings (4.3, 4.4, 4.5)
estudiantes = [
    Estudiante("Ana",   [9.0, 8.5, 9.5]),
    Estudiante("Luis",  [6.0, 5.5, 7.0]),
    Estudiante("María", [10.0, 9.0, 9.5]),
]

for est in estudiantes:
    prom = est.promedio()
    if est.aprobado():
        print(f"✓ {est.nombre}: {prom} -> Aprobado")
    else:
        print(f"✗ {est.nombre}: {prom} -> Reprobado")

# with, as (4.7)
reporte = "\n".join(f"{e.nombre}: {e.promedio()}" for e in estudiantes)
with open("reporte_demo.txt", "w", encoding="utf-8") as archivo:   # as
    archivo.write(reporte)
print("\nReporte guardado en 'reporte_demo.txt'")

# try, except, finally, raise (4.6)
def dividir(a, b):
    if b == 0:
        raise ValueError("No se puede dividir por cero")
    return a / b

try:
    resultado = dividir(10, 0)
except ValueError as e:           # as
    print(f"Error capturado: {e}")
finally:                          # finally
    print("Operación de división intentada (haya error o no).")

# del (4.9)
datos_temporales = [1, 2, 3, 4, 5]
del datos_temporales             # libera la referencia
print("\nFin de la demo.")

Total de palabras reservadas: 35
Todas las keywords: ['False', 'None', 'True', 'and', 'as', 'assert', 'async', 'await', 'break', 'class', 'continue', 'def', 'del', 'elif', 'else', 'except', 'finally', 'for', 'from', 'global', 'if', 'import', 'in', 'is', 'lambda', 'nonlocal', 'not', 'or', 'pass', 'raise', 'return', 'try', 'while', 'with', 'yield']

¿'for' es reservada? True
¿'variable' es reservada? False
¿'class' es reservada? True

No puedes usar palabras reservadas como nombres de variables.

--- Demo: sistema de calificaciones (sin input) ---

✓ Ana: 9.0 -> Aprobado
✗ Luis: 6.17 -> Reprobado
✓ María: 9.5 -> Aprobado

Reporte guardado en 'reporte_demo.txt'
Error capturado: No se puede dividir por cero
Operación de división intentada (haya error o no).

Fin de la demo.


---
## 5. Tipos de Datos Primitivos

Un **tipo de dato** es el conjunto de valores y el conjunto de operaciones definidas en esos valores (Corcuera, IntroPython). Python tiene un gran número de tipos de datos incorporados tales como Números (Integer, Float, Boolean, Complex Number), String, List, Tuple, Set, Dictionary. Otros tipos de datos de alto nivel, tales como Decimal y Fraction, están soportados por módulos externos.

Según la Guía de MoureDev (2026), los **tipos primitivos** fundamentales son:
- **Números enteros**: valores sin decimales, positivos o negativos.
- **Números decimales**: valores con punto flotante, precisión limitada.
- **Booleanos**: verdadero o falso, base de toda lógica.
- **Caracteres**: símbolos individuales de texto (en Python no existe un tipo `char` dedicado).
- **Nulo/Vacío**: ausencia intencional de valor.

### 5.1 `int` — Números Enteros

Tipo de dato (`int`) para representar enteros o números naturales (Corcuera, IntroPython).

**Características clave:**
- A diferencia de C (`int`, `long` con tamaños fijos de 32 o 64 bits), los enteros en Python 3 son de **tamaño ilimitado**.
- Se pueden expresar en diferentes bases:
  - **Decimal** (base 10): `123`
  - **Hexadecimal** con prefijo `0x` o `0X`: `0x1abc`, `0X1ABC`
  - **Octal** con prefijo `0o` o `0O`: `0o1776`
  - **Binario** con prefijo `0b` o `0B`: `0b11000011`

In [5]:
print(123 + 456 - 789)

print(123456789012345678901234567890 + 1)

print(1234567890123456789012345678901234567890 + 1)

resultado = 2 ** 888
print(f"2^888 tiene {len(str(resultado))} dígitos")

print(type(123))

-210
123456789012345678901234567891
1234567890123456789012345678901234567891
2^888 tiene 268 dígitos
<class 'int'>


In [6]:
print(0x1abc)
print(0o1776)
print(0b11000011)

print(hex(1234))
print(bin(254))
print(oct(1234))

6844
1022
195
0x4d2
0b11111110
0o2322


### 5.2 `float` — Números en Punto Flotante

Tipo de dato (`float`) para representar números reales, para uso en aplicaciones científicas o comerciales (Corcuera, IntroPython).

**Características clave:**
- Tienen una representación **IEEE 754 de 64 bits** (típicamente equivalente a `double` en C).
- Tienen **15-17 dígitos decimales de precisión**.
- Para obtener el máximo valor usar `sys.float_info.max`.
- Se puede usar notación científica: `4e5` = 4 × 10⁵.
- No son de tamaño ilimitado como los enteros.

> **Advertencia de precisión:** Los floats son inherentemente imprecisos. Por ejemplo, `0.1 * 3` no da exactamente `0.3` sino `0.30000000000000004`. Para cálculos que requieren precisión exacta (como finanzas), se debe usar el módulo `decimal`.

In [7]:
print(1.23 * -4e5)
print(type(1.2))

import math
print(f"math.pi = {math.pi}")

import random
print(f"random.random() = {random.random()}")

print(f"0.1 * 3 = {0.1 * 3}")

import decimal
x = decimal.Decimal('0.1')
print(f"Decimal('0.1') * 3 = {x * 3}")
print(type(x))

# Error de tamaño de flotante
print(f"1e+1000 = {float(1e1000)}") # Esto devuelve inf (infinito) porque el número es demasiado grande para ser representado como un flotante

-492000.0
<class 'float'>
math.pi = 3.141592653589793
random.random() = 0.6460887670723754
0.1 * 3 = 0.30000000000000004
Decimal('0.1') * 3 = 0.3
<class 'decimal.Decimal'>
1e+1000 = inf


### 5.3 `complex` — Números Complejos

Tipo de dato (`complex`) para representar números complejos de la forma `a + bj` (Corcuera, IntroPython).

**Características clave:**
- La parte imaginaria se denota con `j` (no `i` como en matemáticas).
- Se accede a las partes con `.real` e `.imag`.
- Se puede crear con la función `complex(real, imag)`.
- Soportan operaciones aritméticas directamente.

In [8]:
x = 1 + 2j
print(f"x = {x}")
print(f"Parte real: {x.real}")
print(f"Parte imaginaria: {x.imag}")
print(f"Tipo: {type(x)}")

print(f"x * (3 + 4j) = {x * (3 + 4j)}")

z = complex(2, -3)
print(f"z = {z}")

x = (1+2j)
Parte real: 1.0
Parte imaginaria: 2.0
Tipo: <class 'complex'>
x * (3 + 4j) = (-5+10j)
z = (2-3j)


### 5.4 `bool` — Booleanos

Tipo de dato (`bool`) que tiene solo dos valores: `True` y `False` (Corcuera, IntroPython).

**Características clave:**
- `bool` es una **subclase de `int`**: `True` equivale a `1` y `False` equivale a `0`.
- El entero `0`, un valor vacío (cadena vacía `''`, lista vacía `[]`, tupla vacía `()`, diccionario vacío `{}`), y `None` son tratados como `False`; todo lo demás es `True`.
- Los booleanos se comportan como enteros en operaciones aritméticas.

> **Concepto de "Truthiness" (MoureDev, Guía Junior 2026):** Los booleanos son la base de toda lógica en programación. En Python, cualquier valor puede evaluarse como verdadero o falso en un contexto booleano.

In [9]:
print(8 == 8)
print(8 == 9)
print(type(True))

print(f"bool(0) = {bool(0)}")
print(f"bool(1) = {bool(1)}")
print(f"bool('') = {bool('')}")
print(f"bool('hola') = {bool('hola')}")
print(f"bool([]) = {bool([])}")
print(f"bool([1,2]) = {bool([1,2])}")
print(f"bool(None) = {bool(None)}")

print(f"True + 3 = {True + 3}")
print(f"False + 1 = {False + 1}")
print(f"isinstance(True, int) = {isinstance(True, int)}") # isinstance() te dice si un objeto es una instancia de una clase o tipo específico, en este caso, verifica si True es considerado un entero (int) en Python.

True
False
<class 'bool'>
bool(0) = False
bool(1) = True
bool('') = False
bool('hola') = True
bool([]) = False
bool([1,2]) = True
bool(None) = False
True + 3 = 4
False + 1 = 1
isinstance(True, int) = True


### 5.5 `NoneType` — El valor `None`

Python proporciona un valor especial llamado `None` que representa la **ausencia intencional de valor** (Corcuera, IntroPython).

**Características clave:**
- Equivalente conceptual al puntero `NULL` de C.
- `None` es el único valor del tipo `NoneType`.
- Se usa `is` y `is not` para comprobar si una variable es `None` (no `==`).
- Es común usarlo para inicializar variables o como valor de retorno por defecto de funciones.

In [10]:
x = None
print(f"x = {x}")
print(f"type(x) = {type(x)}")

print(f"x is None: {x is None}")
print(f"x is not None: {x is not None}")

x = None
type(x) = <class 'NoneType'>
x is None: True
x is not None: False


---
## 6. Conversión de Tipos

Se pueden convertir tipos mediante las funciones integradas `int()`, `float()`, `str()`, `bool()`, etc. (Corcuera, IntroPython).

| Función | Convierte a | Ejemplo |
|---------|-------------|--------|
| `int(x)` | Entero | `int('123')` → `123` |
| `float(x)` | Flotante | `float('3.14')` → `3.14` |
| `str(x)` | Cadena | `str(123)` → `'123'` |
| `bool(x)` | Booleano | `bool(0)` → `False` |

**Conversión implícita:** Python convierte automáticamente enteros y flotantes convenientemente. Por ejemplo, `1 + 2.0` produce `3.0` (el `int` se promueve a `float`).

In [11]:
x = '123'
print(f"Original: {x!r}, tipo: {type(x).__name__}")

x = int(x)
print(f"int(x): {x!r}, tipo: {type(x).__name__}")

x = float(x)
print(f"float(x): {x!r}, tipo: {type(x).__name__}")

x = str(x)
print(f"str(x): {x!r}, tipo: {type(x).__name__}")

x = bool(x)
print(f"bool(x): {x!r}, tipo: {type(x).__name__}")

x = str(x)
print(f"str(x): {x!r}, tipo: {type(x).__name__}")

Original: '123', tipo: str
int(x): 123, tipo: int
float(x): 123.0, tipo: float
str(x): '123.0', tipo: str
bool(x): True, tipo: bool
str(x): 'True', tipo: str


In [12]:
# Conversión automática en expresiones mixtas, son deterministas así que no es tipado débil.
print(f"1 + 2.0 = {1 + 2.0}")
print(f"type(1 + 2.0) = {type(1 + 2.0)}")

print(f"True + 5 = {True + 5}")
print(f"False * 10 = {False * 10}")

1 + 2.0 = 3.0
type(1 + 2.0) = <class 'float'>
True + 5 = 6
False * 10 = 0


---
## 7. Operadores

Según la Guía de MoureDev (2026), los operadores son los siguientes:
- **Aritméticos:** `+`, `-`, `*`, `/`, `%` (módulo), `**` (potencia)
- **Comparación:** `==`, `!=`, `<`, `>`, `<=`, `>=`
- **Lógicos:** `and`, `or`, `not` (evaluación de cortocircuito)
- **Asignación:** `=`, `+=`, `-=`, `*=`, `/=` (asignación compuesta)
- **Bit a bit:** `&`, `|`, `^`, `~`, `<<`, `>>` (nivel de bits)

### 7.1 Operadores Aritméticos

Python tiene 7 operadores aritméticos (Bonaretti, Learn Python with Jupyter):

| Operador | Descripción | Ejemplo | Resultado |
|----------|-------------|---------|----------|
| `+` | Suma | `4 + 3` | `7` |
| `-` | Resta | `6 - 2` | `4` |
| `*` | Multiplicación | `6 * 5` | `30` |
| `/` | División flotante | `1 / 2` | `0.5` |
| `//` | División entera (floor) | `7 // 4` | `1` |
| `**` | Exponenciación | `2 ** 5` | `32` |
| `%` | Módulo (residuo) | `9 % 2` | `1` |

**Detalles importantes sobre la división (Corcuera, IntroPython):**
- `/` siempre retorna un `float`: `1 / 2` → `0.5`, `-1 / 2` → `-0.5`
- `//` retorna el piso (floor integer): `1 // 2` → `0`, `-1 // 2` → `-1`
- `//` con floats: `8.9 // 2.5` → `3.0`, `-8.9 // 2.5` → `-4.0`
- `%` con negativos: `-9 % 2` → `1`, `9 % -2` → `-1`

In [13]:
print("--- División flotante ---")
print(f"1 / 2 = {1 / 2}")
print(f"-1 / 2 = {-1 / 2}")
print(f"10 / 5 = {10 / 5}")

print("\n--- División entera (floor) ---")
print(f"1 // 2 = {1 // 2}")
print(f"-1 // 2 = {-1 // 2}")
print(f"7 // 4 = {7 // 4}")
print(f"8.9 // 2.5 = {8.9 // 2.5}")
print(f"-8.9 // 2.5 = {-8.9 // 2.5}")

print("\n--- Módulo ---")
print(f"9 % 2 = {9 % 2}")
print(f"-9 % 2 = {-9 % 2}")
print(f"9 % -2 = {9 % -2}")
print(f"7.5 % 4 = {7.5 % 4}")

print("\n--- Exponenciación ---")
print(f"2 ** 5 = {2 ** 5}")
print(f"1.2 ** 3.4 = {1.2 ** 3.4}")

--- División flotante ---
1 / 2 = 0.5
-1 / 2 = -0.5
10 / 5 = 2.0

--- División entera (floor) ---
1 // 2 = 0
-1 // 2 = -1
7 // 4 = 1
8.9 // 2.5 = 3.0
-8.9 // 2.5 = -4.0

--- Módulo ---
9 % 2 = 1
-9 % 2 = 1
9 % -2 = -1
7.5 % 4 = 3.5

--- Exponenciación ---
2 ** 5 = 32
1.2 ** 3.4 = 1.858729691979481


### Funciones integradas para números

Python contiene funciones integradas para manipular números (Corcuera, IntroPython):
- **Matemáticas:** `round()`, `pow()`, `abs()`
- **Conversión de tipos:** `int()`, `float()`, `str()`, `bool()`, `type()`
- **Conversión de base:** `hex()`, `bin()`, `oct()`

In [14]:
x = 1.23456

print(f"round({x}) = {round(x)}")
print(f"round({x}, 1) = {round(x, 1)}")
print(f"round({x}, 2) = {round(x, 2)}")

print(f"pow(2, 5) = {pow(2, 5)}")
print(f"abs(-4.1) = {abs(-4.1)}")

print(f"hex(1234) = {hex(1234)}")
print(f"bin(254) = {bin(254)}")
print(f"oct(1234) = {oct(1234)}")
print(f"0xABCD = {0xABCD}")

round(1.23456) = 1
round(1.23456, 1) = 1.2
round(1.23456, 2) = 1.23
pow(2, 5) = 32
abs(-4.1) = 4.1
hex(1234) = 0x4d2
bin(254) = 0b11111110
oct(1234) = 0o2322
0xABCD = 43981


### 7.2 Operadores de Comparación

Los operadores de comparación se aplican a enteros y flotantes y producen un resultado booleano (Corcuera, IntroPython):

| Operador | Descripción | Ejemplo |
|----------|-------------|--------|
| `<` | Menor que | `2 < 13` → `True` |
| `<=` | Menor o igual que | `2 <= 2` → `True` |
| `>` | Mayor que | `13 > 2` → `True` |
| `>=` | Mayor o igual que | `3 >= 3` → `True` |
| `==` | Igual a | `2 == 3` → `False` |
| `!=` | Diferente de | `3 != 2` → `True` |
| `in` | Contenido en secuencia | `4 in [1,4,3]` → `True` |
| `not in` | No contenido en secuencia | `6 not in [1,4,3]` → `True` |
| `is` | Mismo objeto en memoria | `type(x) is int` |
| `is not` | Diferente objeto en memoria | `type(x) is not int` |

> **Puente con C:** En C, la comparación de igualdad usa `==` igual que Python, pero `is` no tiene equivalente directo. En Python `==` compara **valores** mientras que `is` compara **identidad** (si apuntan al mismo objeto en memoria).

In [15]:
number = 5

print(f"number > 3 = {number > 3}")
print(f"number < 3 = {number < 3}")
print(f"number == 5 = {number == 5}")
print(f"number != 5 = {number != 5}")

result = number > 3
print(f"result = {result}, type = {type(result)}")

lis = [1, 4, 3, 2, 5]
print(f"4 in lis = {4 in lis}")
print(f"6 not in lis = {6 not in lis}")

x = 5
print(f"type(x) is int = {type(x) is int}")

number > 3 = True
number < 3 = False
number == 5 = True
number != 5 = False
result = True, type = <class 'bool'>
4 in lis = True
6 not in lis = True
type(x) is int = True


### 7.3 Operadores Lógicos

A diferencia de C (`&&`, `||`, `!`), Python usa palabras legibles (Corcuera, IntroPython):

| Operador | Descripción | Equivalente en C |
|----------|-------------|-----------------|
| `and` | AND lógico | `&&` |
| `or` | OR lógico | `\|\|` |
| `not` | NOT lógico | `!` |

**Evaluación de cortocircuito (MoureDev, Guía Junior 2026):**
- `and`: si el primer operando es `False`, no evalúa el segundo.
- `or`: si el primer operando es `True`, no evalúa el segundo.

In [16]:
number = 3

print("--- Tabla AND ---")
print(f"True and True = {True and True}")
print(f"True and False = {True and False}")
print(f"False and True = {False and True}")
print(f"False and False = {False and False}")

print("\n--- Combinando condiciones ---")
print(f"number > 1 = {number > 1}")
print(f"number < 5 = {number < 5}")
print(f"number > 1 and number < 5 = {number > 1 and number < 5}")

print(f"\nTrue and not False = {True and not False}")

--- Tabla AND ---
True and True = True
True and False = False
False and True = False
False and False = False

--- Combinando condiciones ---
number > 1 = True
number < 5 = True
number > 1 and number < 5 = True

True and not False = True


### 7.4 Operadores de Bits (Bitwise)

Permiten operaciones a nivel de bits. Funcionan igual que en C (Corcuera, IntroPython):

| Operador | Descripción | Ejemplo |
|----------|-------------|--------|
| `&` | AND bit a bit | `0b10000001 & 0b10001111` → `0b10000001` |
| `\|` | OR bit a bit | `0b10000001 \| 0b10001111` → `0b10001111` |
| `~` | NOT bit a bit | `~0b10000001` → `-0b10000010` |
| `^` | XOR bit a bit | `0b10000001 ^ 0b10001111` → `0b00001110` |
| `<<` | Desplazamiento izquierda | `0b10000001 << 2` → `0b1000000100` |
| `>>` | Desplazamiento derecha | `0b10000001 >> 2` → `0b100000` |

**Nota:**
El operador `~` (tilde) no realiza una simple inversión estética de los dígitos visibles en la pantalla, sino que se define rigurosamente mediante la siguiente relación:

$$\sim x = -x - 1$$

Esto debido a superar las dificultades de representar los números negativos junto al cero.
Si se toma un byte común (8 bits) para evitar el infinito de Python, el número más alto con el bit de signo en cero (positivo) es el 127 (01111111). El número 0 se escribe 00000000. Si aplicamos ~ al cero (00000000), el sistema invierte todos los bits y obtenemos 11111111. En la convención de complemento a dos, la secuencia 11111111 no representa el número más alto, sino el $-1$. De este modo, se cumple la fórmula:$$\sim 0 = -0 - 1 = -1$$El desplazamiento de una unidad ocurre porque el sistema debe dar cabida al cero dentro de la escala numérica sin duplicar su signo —un inconveniente crónico de sistemas más primitivos como el signo-magnitud, donde existían el $+0$ y el $-0$—, pero forzando a que la inversión pura de bits se sitúe siempre un paso por detrás del valor negativo directo.

In [17]:
x = 0b10000001
y = 0b10001111
z = 0b00000000

print(f"x = {bin(x)} ({x})")
print(f"y = {bin(y)} ({y})")
print(f"z = {bin(z)} ({z})")
print(f"x & y  = {bin(x & y)} ({x & y})")
print(f"x | y  = {bin(x | y)} ({x | y})")
print(f"~x     = {bin(~x)} ({~x})") 
print(f"~y     = {bin(~y)} ({~y})")
print(f"~z     = {bin(~z)} ({~z})")
print(f"x ^ y  = {bin(x ^ y)} ({x ^ y})")
print(f"x << 2 = {bin(x << 2)} ({x << 2})")
print(f"x >> 2 = {bin(x >> 2)} ({x >> 2})")

print(f"\n--- Ejemplo simple: 5 & 3 ---")
print(f"5 = {bin(5)} (0101)")
print(f"3 = {bin(3)} (0011)")
print(f"5 & 3 = {5 & 3} ({bin(5 & 3)}) = 0001")

x = 0b10000001 (129)
y = 0b10001111 (143)
z = 0b0 (0)
x & y  = 0b10000001 (129)
x | y  = 0b10001111 (143)
~x     = -0b10000010 (-130)
~y     = -0b10010000 (-144)
~z     = -0b1 (-1)
x ^ y  = 0b1110 (14)
x << 2 = 0b1000000100 (516)
x >> 2 = 0b100000 (32)

--- Ejemplo simple: 5 & 3 ---
5 = 0b101 (0101)
3 = 0b11 (0011)
5 & 3 = 1 (0b1) = 0001


### 7.5 Operadores de Asignación Compuesta

Permiten combinar una operación con la asignación en un solo paso (Corcuera, IntroPython):

| Operador | Ejemplo | Equivalente a |
|----------|---------|---------------|
| `=` | `x = 5` | `x = 5` |
| `+=` | `x += 5` | `x = x + 5` |
| `-=` | `x -= 5` | `x = x - 5` |
| `*=` | `x *= 5` | `x = x * 5` |
| `/=` | `x /= 5` | `x = x / 5` |
| `%=` | `x %= 5` | `x = x % 5` |
| `//=` | `x //= 5` | `x = x // 5` |
| `**=` | `x **= 5` | `x = x ** 5` |
| `&=` | `x &= 5` | `x = x & 5` |
| `\|=` | `x \|= 5` | `x = x \| 5` |
| `^=` | `x ^= 5` | `x = x ^ 5` |
| `>>=` | `x >>= 5` | `x = x >> 5` |
| `<<=` | `x <<= 5` | `x = x << 5` |

In [18]:
x = 10
print(f"x = {x}")

x += 5
print(f"x += 5  → x = {x}")

x -= 3
print(f"x -= 3  → x = {x}")

x *= 2
print(f"x *= 2  → x = {x}")

x //= 3
print(f"x //= 3 → x = {x}")

x **= 2
print(f"x **= 2 → x = {x}")

x = 10
x += 5  → x = 15
x -= 3  → x = 12
x *= 2  → x = 24
x //= 3 → x = 8
x **= 2 → x = 64


---
## 8. Cadenas de Texto (Strings)

Tipo de dato (`str`) para representar cadenas de caracteres, para uso en procesado de textos (Corcuera, IntroPython).

**Características fundamentales:**
- Se delimitan por `'...'`, `"..."`, `'''...'''`, o `"""..."""`.
- Python 3 usa el conjunto de caracteres **Unicode**.
- Los strings son **inmutables**: su contenido no se puede modificar directamente.
- Para convertir números en strings se usa `str()`.
- Para convertir strings a números se usa `int()` o `float()`.
- Python **no tiene un tipo `char` dedicado**: un carácter es simplemente un string de longitud 1.

> **Puente con C:** En C, las cadenas son arrays de `char` terminados en `\0` y son mutables (`str[0] = 'A'`). En Python, los strings son objetos inmutables. Intentar `nombre[0] = 'E'` genera un `TypeError`.

### 8.1 Creación de Strings y Secuencias de Escape

Para especificar caracteres especiales se usan **secuencias de escape**:

| Secuencia | Significado |
|-----------|-------------|
| `\n` | Salto de línea |
| `\t` | Tabulación |
| `\r` | Retorno de carro |
| `\\` | Barra invertida |
| `\'` | Comilla simple |
| `\"` | Comilla doble |

In [19]:
s1 = 'apple'
print(f"s1 = {s1!r}")

s2 = "orange"
print(f"s2 = {s2!r}")

s3 = "'orange'"
print(f"s3 = {s3!r}")

s3 = "\"orange\""
print(f"s3 = {s3!r}")

s4 = """testing
testing"""
print(f"s4 = {s4!r}")

mensaje = "Hola\n\tMundo"
print(mensaje)

s1 = 'apple'
s2 = 'orange'
s3 = "'orange'"
s3 = '"orange"'
s4 = 'testing\ntesting'
Hola
	Mundo


### 8.2 Funciones y Operadores para Strings

| Función/Operador | Descripción | Ejemplo |
|------------------|-------------|--------|
| `len()` | Longitud | `len('Hello')` → `5` |
| `in` | ¿Contenido? | `'ell' in 'Hello'` → `True` |
| `+` | Concatenación | `'Hello' + '!'` → `'Hello!'` |
| `*` | Repetición | `'Hello' * 2` → `'HelloHello'` |
| `[i]`, `[-i]` | Indexación | `'Hello'[1]` → `'e'` |
| `[m:n:step]` | Slicing | `'Hello'[1:3]` → `'el'` |

In [20]:
s = "Hello, world"

print(f"type(s) = {type(s)}")
print(f"len(s) = {len(s)}")
print(f"'ello' in s = {'ello' in s}")

print(f"\n--- Indexación ---")
print(f"s[0] = {s[0]!r}")
print(f"s[1] = {s[1]!r}")
print(f"s[-1] = {s[-1]!r}")
print(f"s[-2] = {s[-2]!r}")

print(f"\n--- Slicing ---")
print(f"s[1:3] = {s[1:3]!r}")
print(f"s[1:-1] = {s[1:-1]!r}")
print(f"s[:4] = {s[:4]!r}")
print(f"s[4:] = {s[4:]!r}")
print(f"s[:] = {s[:]!r}")
print(f"s[0:5:2] = {s[0:5:2]!r}")

print(f"\n--- Concatenación y Repetición ---")
s2 = s + " again"
print(f"s + ' again' = {s2!r}")
print(f"s2 * 2 = {(s2 * 2)!r}")

type(s) = <class 'str'>
len(s) = 12
'ello' in s = True

--- Indexación ---
s[0] = 'H'
s[1] = 'e'
s[-1] = 'd'
s[-2] = 'l'

--- Slicing ---
s[1:3] = 'el'
s[1:-1] = 'ello, worl'
s[:4] = 'Hell'
s[4:] = 'o, world'
s[:] = 'Hello, world'
s[0:5:2] = 'Hlo'

--- Concatenación y Repetición ---
s + ' again' = 'Hello, world again'
s2 * 2 = 'Hello, world againHello, world again'


### 8.3 Inmutabilidad de Strings

Los strings son **inmutables**: no puedes cambiar un carácter individual como harías en un array de caracteres de C (`char str[]`).

In [21]:
nombre = "Ada Lovelace"

# nombre[0] = 'E'  # TypeError: 'str' object does not support item assignment

nombre_modificado = "E" + nombre[1:]
print(f"Original: {nombre}")
print(f"Modificado: {nombre_modificado}")

Original: Ada Lovelace
Modificado: Eda Lovelace


### 8.4 Métodos de Strings

La clase `str` proporciona varios métodos útiles (Corcuera, IntroPython):

| Método | Descripción |
|--------|-------------|
| `s.strip()` | Elimina espacios al inicio y final |
| `s.rstrip()` | Elimina espacios al final |
| `s.lstrip()` | Elimina espacios al inicio |
| `s.upper()` | Convierte a mayúsculas |
| `s.lower()` | Convierte a minúsculas |
| `s.isupper()` | ¿Es todo mayúsculas? |
| `s.islower()` | ¿Es todo minúsculas? |
| `s.find(sub)` | Busca subcadena, retorna índice o -1 |
| `s.index(sub)` | Busca subcadena, retorna índice (error si no existe) |
| `s.startswith(prefix)` | ¿Empieza con...? |
| `s.endswith(suffix)` | ¿Termina con...? |
| `s.split(sep)` | Divide en lista por separador |
| `sep.join(list)` | Une lista de strings con separador |

In [22]:
texto = "  Hola Mundo Python  "

print(f"Original: {texto!r}")
print(f"strip(): {texto.strip()!r}")
print(f"lstrip(): {texto.lstrip()!r}")
print(f"rstrip(): {texto.rstrip()!r}")

s = "Hello, World"
print(f"\nupper(): {s.upper()!r}")
print(f"lower(): {s.lower()!r}")
print(f"isupper(): {s.isupper()}")

print(f"\nfind('World'): {s.find('World')}")
print(f"find('xyz'): {s.find('xyz')}")
print(f"startswith('Hello'): {s.startswith('Hello')}")
print(f"endswith('World'): {s.endswith('World')}")

print(f"\nsplit(', '): {s.split(', ')}")

palabras = ['Python', 'es', 'genial']
print(f"' '.join(palabras): {' '.join(palabras)!r}")

Original: '  Hola Mundo Python  '
strip(): 'Hola Mundo Python'
lstrip(): 'Hola Mundo Python  '
rstrip(): '  Hola Mundo Python'

upper(): 'HELLO, WORLD'
lower(): 'hello, world'
isupper(): False

find('World'): 7
find('xyz'): -1
startswith('Hello'): True
endswith('World'): True

split(', '): ['Hello', 'World']
' '.join(palabras): 'Python es genial'


### 8.5 ¿Tipo carácter? `ord()` y `chr()`

Python no tiene un tipo de dato dedicado a caracteres. Un carácter es un string de longitud 1. Las funciones integradas `ord()` y `chr()` operan sobre strings de longitud 1 (Corcuera, IntroPython):

- `ord(c)`: retorna el ordinal Unicode (entero) del carácter.
- `chr(i)`: retorna el carácter correspondiente al ordinal Unicode.

In [23]:
print(f"ord('A') = {ord('A')}")
print(f"ord('水') = {ord('水')}")
print(f"chr(65) = {chr(65)!r}")
print(f"chr(27700) = {chr(27700)!r}")

ord('A') = 65
ord('水') = 27700
chr(65) = 'A'
chr(27700) = '水'


### 8.6 Formato de Strings

Existen varias formas de formatear strings en Python (Corcuera, IntroPython):

1. **f-strings** (Python 3.6+, la más moderna y recomendada)
2. **`str.format()`** (Python 3.x)
3. **Operador `%`** (estilo C, legacy)

#### f-strings (Format Strings)

La forma más moderna y recomendada. Permiten insertar expresiones directamente dentro del string usando `{}`.

In [4]:
nombre = "Ada Lovelace"
edad = 36
precio = 14.5678
numero_grande = 1_000_000_000

print(f"La programadora {nombre} vivió {edad} años.")
print(f"El precio formateado a 2 decimales es: {precio:.2f}")
print(f"2 + 3 = {2 + 3}")
print(f"{'Hola':>10} | {'Mundo':<10} | {'Python':^10}")
print(f"Numero_grande normal = {numero_grande}")
print(f"numero_grande = {numero_grande:,}")  # Formato con separador de miles
print(f"numero_grande en hexadecimal = {numero_grande:#x}")
print(f"numero_grande en binario = {numero_grande:#b}")

La programadora Ada Lovelace vivió 36 años.
El precio formateado a 2 decimales es: 14.57
2 + 3 = 5
      Hola | Mundo      |   Python  
Numero_grande normal = 1000000000
numero_grande = 1,000,000,000
numero_grande en hexadecimal = 0x3b9aca00
numero_grande en binario = 0b111011100110101100101000000000


#### `str.format()`

Usa la función `format()` y `{}` como campos de reemplazo (Corcuera, IntroPython):

In [25]:
print('|{}|{}|more|'.format('Hello', 'world'))

print('|{0}|{1}|more|'.format('Hello', 'world'))
print('|{1}|{0}|more|'.format('Hello', 'world'))

print('|{greeting}|{name}|'.format(greeting='Hello', name='Peter'))

print('|{1:8}|{0:>7}|'.format('Hello', 'Peter'))

print('|{a:.3f}|{b:6.2f}|{c:4d}|'.format(a=1.2, b=3.456, c=78))

|Hello|world|more|
|Hello|world|more|
|world|Hello|more|
|Hello|Peter|
|Peter   |  Hello|
|1.200|  3.46|  78|


#### Alineación con métodos de string

Se pueden usar las funciones `str.rjust(n)`, `str.ljust(n)`, `str.center(n)`, `str.zfill(n)` donde `n` es el ancho de campo (Corcuera, IntroPython):

In [26]:
print(f"'123'.rjust(5) = {'123'.rjust(5)!r}")
print(f"'123'.ljust(5) = {'123'.ljust(5)!r}")
print(f"'123'.center(5) = {'123'.center(5)!r}")
print(f"'123'.zfill(5) = {'123'.zfill(5)!r}")
print(f"'-1.2'.zfill(6) = {'-1.2'.zfill(6)!r}")

'123'.rjust(5) = '  123'
'123'.ljust(5) = '123  '
'123'.center(5) = ' 123 '
'123'.zfill(5) = '00123'
'-1.2'.zfill(6) = '-001.2'


---
## 9. Ejercicio Práctico: Calculadora de Calificaciones

**Contexto:** Estás en la escuela (México) y quieres saber si ya tienes suficiente para pasar la materia. La nota máxima es 10 y la mínima aprobatoria es 6.

Cada parcial tiene sus propios porcentajes para tareas y examen:
- **Parcial 1:** 60 % tareas, 40 % examen
- **Parcial 2:** 60 % tareas, 40 % examen
- **Parcial 3:** 20 % tareas, 80 % examen

**Instrucciones:**
1. Define variables para las calificaciones de cada componente en los 3 parciales. Usa estos valores de ejemplo:
   - Parcial 1: tareas = 8.5, examen = 7.0
   - Parcial 2: tareas = 9.0, examen = 6.5
   - Parcial 3: tareas = 7.0, examen = 8.0
2. Calcula la calificación de **cada parcial** según sus ponderaciones (tareas% + examen%).
3. Calcula el **promedio final** de los 3 parciales.
4. Usa `f-strings` para imprimir un reporte con la calificación de cada parcial y el promedio final formateado a 2 decimales.

**Restricciones:**
- Aún no hemos visto bucles (`for`/`while`) ni listas (`list`), por lo que debes crear variables individuales para cada valor.

**Conceptos que aplicarás:**
- Variables y asignación
- Operadores aritméticos (`*`, `+`, `/`)
- f-strings para formateo de salida

In [27]:
# Escribe tu solución aquí:

# Parcial 1 — 60 % tareas, 40 % examen
tareas_1 = 8.5
examen_1 = 7.0

# Parcial 2 — 60 % tareas, 40 % examen
tareas_2 = 9.0
examen_2 = 6.5

# Parcial 3 — 20 % tareas, 80 % examen
tareas_3 = 7.0
examen_3 = 8.0

# Calcula el promedio final

# Imprime el reporte